### PR TITLE
[Tools] Query Features

### PR DESCRIPTION
**Pull request by 0xzoz**
- Add QueryTypes(BlockHeight, Resources, MoveValue)
- Refactor and make output structure conform

### CODE REVIEW

File: tools/query/src/get_client.rs [hunk 1]
```diff
@@ -1,30 +1,39 @@
...
-  //     })
-  // ).await?;
-  
-  let nodes = &app_cfg.profile.upstream_nodes;
-  let url = nodes.iter().next().context("cannot get url")?;
-  let client = Client::new(url.to_owned());
-  let res = client.get_index().await?;
-  
-  Ok((client, ChainId::new(res.inner().chain_id)))
+pub async fn get_local_node() -> anyhow::Result<(Client, ChainId)> {
```
Comment (0o-de-lally) [inline tools/query/src/get_client.rs:29]: This we can keep!
I would move it to clien_ext.rs as a constructor. Like this:
https://github.com/0o-de-lally/libra-framework/blob/fee-payer-patch/types/src/type_extensions/client_ext.rs#L79

File: tools/query/src/query_cli.rs [hunk 2]
```diff
@@ -17,10 +20,16 @@ impl QueryCli {
...
         // TODO: get client from configs
 
-        let res = self.subcommand.query_to_json(None).await?;
-
-        println!("{}", res);
+        let output = self.subcommand.query(None).await?;
 
+        match output {
+            OutputType::Json(json_str) => {
+                colorize_and_print(&json_str)?;
```
Comment (0o-de-lally) [inline tools/query/src/query_cli.rs:27]: colors. evil. but ok if you insist ;)

File: tools/query/src/query_type.rs [hunk 3]
```diff
@@ -1,14 +1,21 @@
 use crate::{
```
Comment (0o-de-lally) [inline tools/query/src/query_type.rs:1]: Everything looks good in this file. Well done!

File: tools/query/src/query_type.rs [hunk 4]
```diff
@@ -96,84 +103,264 @@ pub enum QueryType {
...
+            }
+            QueryType::LookupAddress { auth_key } => {
+                let addr = lookup_originating_address(&client, auth_key.to_owned()).await?;
 
-    }
+                Ok(OutputType::Json(serde_json::to_string(&json!({
+                    "address": addr
+                }))?))
+            }
+            QueryType::MoveValue {
```
Comment (0o-de-lally) [inline tools/query/src/query_type.rs:229]: I would first move this to it's own file.
Also I'm sure BCS can do a lot of this conversion.  Look for examples in Diem of `let x: SomeMoveValue = BCS:from_bcs(resource);

File: tools/query/src/query_view.rs [hunk 5]
```diff
@@ -12,16 +12,26 @@ pub async fn get_view(
...
-pub fn display_view(res: Vec<Value>) -> Result<String> {
-    let values_to_string = res.iter().map(|v| v.to_string()).collect::<Vec<_>>();
-    if values_to_string.len() > 1 {
-        Ok(format!("[{}]", values_to_string.join(", ")))
-    } else {
-        Ok(format!(
-            "[{}]",
-            values_to_string.first().expect("api didn't return a value")
-        ))
+pub async fn fetch_and_display(
```
Comment (0o-de-lally) [inline tools/query/src/query_view.rs:15]: I would keep display_view() as a separate function.
If you want this to be a one step then I would call  out to that function.

File: tools/query/src/query_view.rs [hunk 6]
```diff
@@ -12,16 +12,26 @@ pub async fn get_view(
...
+        // If the array has only one item, return it under the derived key.
+        if arr.len() == 1 {
+            return Ok(json!({ key: &arr[0] }));
+        }
+
+        // If the array has more than one item, return the entire array under the derived key.
+        return Ok(json!({ key: arr }));
     }
+
+    Ok(Value::String("Unable to parse response".to_string()))
```
Comment (0o-de-lally) [inline tools/query/src/query_view.rs:36]: should this be an Error? bail!() should do it here.

File: tools/query/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
+use anyhow::Context;
+use colored::Colorize;
+use serde_json::Value;
+
+pub fn colorize_and_print(json_str: &str) -> anyhow::Result<()> {
```
Comment (0o-de-lally) [inline tools/query/src/utils.rs:5]: nice work! I would checkout Diem and there is a CliTypedResult, or such, which does something similar. without colors ;)

**Review by 0o-de-lally — CHANGES_REQUESTED** [review]
Very impressive lift here. Level up! Some small considerations below...

File: tools/query/src/get_client.rs [hunk 8]
```diff
@@ -1,30 +1,39 @@
...
+use url::Url;
 use zapatos_sdk::rest_client::Client;
 use zapatos_sdk::types::chain_id::ChainId;
 
-pub async fn find_good_upstream(app_cfg: &AppCfg) -> anyhow::Result<(Client, ChainId)> {
-    // check if we can connect to this client, or exit
+const LOCAL_NODE_URL: &str = "http://localhost:8080";
 
-  // TODO: iterate through all and find a valid one.
+pub fn get_libra_config_path() -> PathBuf {
```
[the remark this answers is not in the document]
Comment (0xzoz) [reply tools/query/src/get_client.rs]: addressed in [9fd79d1](https://github.com/0LNetworkCommunity/libra-framework/pull/27/commits/9fd79d1c5a82e35849a1bb49179faa9d6a4917ce)

File: tools/query/src/get_client.rs [hunk 9]
```diff
@@ -1,30 +1,39 @@
...
+    let mut config_path = dirs::home_dir().expect("Failed to get home directory");
+    config_path.push(".libra");
+    config_path.push("libra.yaml");
+    config_path
+}
+
+pub async fn find_good_upstream(app_cfg: &AppCfg) -> anyhow::Result<(Client, ChainId)> {
```
[the remark this answers is not in the document]
Comment (0xzoz) [reply tools/query/src/get_client.rs]: addressed in [361c448](https://github.com/0LNetworkCommunity/libra-framework/pull/27/commits/361c4481ffae1a2337cc12cd96466974823fad96)

File: tools/query/src/get_client.rs (hunk 1 above)
Comment (0xzoz) [reply tools/query/src/get_client.rs]: addressed in [a87d5a1](https://github.com/0LNetworkCommunity/libra-framework/pull/27/commits/a87d5a144e599c059b255da10c022bbae3738b8d)

**Comment by 0xzoz** [comment]
Removing this as its been superseded by #47

**Pull request closed without merging**